lib/active_merchant/billing/gateways/adyen.rb

```diff
@@ -459,6 +466,23 @@ def add_browser_info(browser_info, post)
           userAgent: browser_info[:user_agent]
         }
       end
+
+      def failure_response(initial_response)
+        Response.new(
+          false,
+          'Recurring transactions are not supported for this card type.',
+          initial_response.params,
+          authorization: initial_response.authorization,
+          test: initial_response.test,
+          error_code: initial_response.error_code,
+          avs_result: initial_response.avs_result,
+          cvv_result: initial_response.cvv_result[:code]
+        )
+      end
+
+      def card_not_stored?(response)
+        response.authorization.split('#')[2].nil?
```

Comment:
Is this null-safe? In other words, is it always guaranteed that ```response.authorization``` will be a string when this line executes? I think this question translates to, will the ```pspReference``` in the initial response always be present? (Traces down to here: https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/adyen.rb#L473). If not or if it depends on what the remote server does, it would be best to add a null check to be safe.

So something like ```response.authorization && response.authorization.split('#')[2].nil?``` if that conditional retains the implication of the card not being stored.

Reply:
From what I've seen `pspReference` always exists; that seems to be the transaction ID. It's just the `recurring.recurringDetailReference` that doesn't always get returned. So from what I know and what I've seen, this should be fine. Is there some other check that I can/should do to confirm that though?

Reply:
I would just check for null (like in ```response.authorization && ...```). That would make the code more robust, preventing an error in the case that the gateway does something different, someone uses the method for some other case where the authorization could be ```nil```, or that it's possible here for some reason we're unaware of. Looking at ```authorization_from``` again, I realized that the early return of ```nil`` ` in the case of ```pspReference``` not being present is itself a null check. So, we'd actually just be cascading that null check to this method.

Reply:
I went with a ternary to check to see if `response.authorization` exists that returns `true`, i.e., the card was not successfully stored, if it doesn't exist.

`response.authorization ? response.authorization.split('#')[2].nil? : true`